ambry-api/src/main/java/com.github.ambry/config/CryptoServiceConfig.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.config;
+
+/**
+ * Configuration parameters required by a {@link com.github.ambry.router.CryptoService}.
+ * <p/>
+ * Receives the in-memory representation of a properties file and extracts parameters that are specifically
+ * required for {@link com.github.ambry.router.CryptoService} and presents them for retrieval through defined APIs.
+ */
+public class CryptoServiceConfig {
+
+  /**
+   * The mode to encrypt or decrypt. Accepted values are "GCM" for now.
+   * Ensure this algo goes hand in hand with {@link com.github.ambry.router.KeyManagementService} implementation
```

Comment:
I think there could be some issues with the generics when we integrate this with router. If the router takes in the type erased `CryptoService` and `KeyManagementService` we won't get any compile time checking for the key type. Alternatively, if it takes `CryptoService<?>` and `KeyManagementService<?>`, it won't pass type checking when you pass a key from KMS to CS. 

A solution could be to parameterize `NonBlockingRouter` as `NonBlockingRouter<KeyType>` and use that to type check the arguments passed into the constructor. Getting the factory via reflection could potentially not check the generic type though (https://en.wikipedia.org/wiki/Generics_in_Java#Problems_with_type_erasure)

I guess you could also have CryptoService completely encapsulate the key management and provide a `encrypt/decrypt(buffer, aid, cid)` method.